core/libp2p/muxer/yamux/yamuxed_connection.hpp

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_YAMUXED_CONNECTION_IMPL_HPP
+#define KAGOME_YAMUXED_CONNECTION_IMPL_HPP
+
+#include <functional>
+#include <map>
+#include <queue>
+
+#include <boost/asio/streambuf.hpp>
+#include <boost/system/error_code.hpp>
+#include "common/buffer.hpp"
+#include "common/logger.hpp"
+#include "libp2p/muxer/muxer_adaptor.hpp"
+#include "libp2p/muxer/yamux/yamux_config.hpp"
+#include "libp2p/muxer/yamux/yamux_stream_parameters.hpp"
+#include "libp2p/stream/stream.hpp"
+#include "libp2p/transport/connection.hpp"
+#include "libp2p/transport/muxed_connection.hpp"
+#include "libp2p/connection/stream.hpp"
+
+namespace libp2p::stream {
+  class YamuxStream;
+}
+
+namespace libp2p::muxer {
+  struct YamuxFrame;
+
+  /**
+   * Implementation of multiplexed connection - connection, which has only one
+   * physical link to another peer, but many logical streams, for example, for
+   * several applications
+   * Read more: https://github.com/hashicorp/yamux/blob/master/spec.md
+   */
+  class YamuxedConnection
+      : public transport::MuxedConnection,
+        public std::enable_shared_from_this<YamuxedConnection> {
+   public:
+    using StreamId = uint32_t;
+
+    enum class YamuxErrorStream {
+      NO_SUCH_STREAM = 1,
+      NOT_WRITABLE,
+      NOT_READABLE,
+      YAMUX_IS_CLOSED
+    };
+
+    /**
+     * Create a new Yamux instance
+     * @param connection, multiplexed by this instance
+     * @param stream_handler - function, which is going to be called, when a new
+     * stream arrives
+     * @param yamux_config to configure this instance
+     */
+    YamuxedConnection(std::shared_ptr<transport::Connection> connection,
+                      std::function<connection::Stream::Handler> stream_handler,
+                      YamuxConfig yamux_config,
+                      kagome::common::Logger logger =
+                          kagome::common::createLogger("YamuxedConnection"));
+
+    YamuxedConnection(const YamuxedConnection &other) = delete;
```

Comment:
maybe better replace by inheriting from boost::noncopyable or similar?

Reply:
It is yamux impl, I did not touch it (I don't know why it is here though)